core/src/main/java/org/elasticsearch/search/aggregations/metrics/percentiles/ParsedPercentiles.java

```diff
@@ -0,0 +1,187 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.metrics.percentiles;
+
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.ParsedAggregation;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+public class ParsedPercentiles extends ParsedAggregation implements Percentiles {
+
+    private final String type;
+
+    // Map of <Key, Tuple<Value, Value As String>>
+    protected final Map<Double, Tuple<Double, String>> percentiles = new HashMap<>();
+    protected boolean keyed;
+
+    protected ParsedPercentiles(String type) {
+        this.type = type;
+    }
+
+    private Tuple<Double, String> getPercentile(double percent) {
+        return percentiles.get(percent);
+    }
+
+    @Override
+    public double percentile(double percent) {
+        Tuple<Double, String> percentile = getPercentile(percent);
+        return percentile != null ? percentile.v1() : 0;
+    }
+
+    @Override
+    public String percentileAsString(double percent) {
+        Tuple<Double, String> percentile = getPercentile(percent);
+        return percentile != null ? percentile.v2() : null;
+    }
+
+    @Override
+    public Iterator<Percentile> iterator() {
+        return new Iterator<Percentile>() {
+            final Iterator<Map.Entry<Double, Tuple<Double, String>>> iterator = percentiles.entrySet().iterator();
+            @Override
+            public boolean hasNext() {
+                return iterator.hasNext();
+            }
+
+            @Override
+            public Percentile next() {
+                Map.Entry<Double, Tuple<Double, String>> next = iterator.next();
+                return new InternalPercentile(next.getKey(), next.getValue().v1());
+            }
+        };
+    }
+
+    @Override
+    protected String getType() {
+        return type;
+    }
+
+    @Override
+    protected XContentBuilder doXContentBody(XContentBuilder builder, Params params) throws IOException {
+        if (keyed) {
+            builder.startObject(InternalAggregation.CommonFields.VALUES.getPreferredName());
+            for (Map.Entry<Double, Tuple<Double, String>> percentile : percentiles.entrySet()) {
+                builder.field(String.valueOf(percentile.getKey()), percentile.getValue().v1());
+
+                String valueAsString = percentile.getValue().v2();
+                if (valueAsString != null && valueAsString.isEmpty() == false) {
+                    builder.field(percentile.getKey() + "_as_string", percentile.getValue().v2());
```

Comment:
Yuck! I think I see now why these response keys make using ObjectParser almost impossible. Nothing to do about this unfortunately, I think.